src/Bicep.Core/Parser/Parser.cs

```diff
@@ -297,25 +313,51 @@ private SyntaxBase ParenthesizedExpression()
 
         private SyntaxBase FunctionCallOrVariableAccess()
         {
-            var identifier = this.Identifier(b => b.ExpectedFunctionName());
+            var identifier = this.Identifier(b => b.ExpectedVariableOrFunctionName());
 
-            if (this.Check(TokenType.LeftParen) == false)
+            if (Check(TokenType.LeftParen))
             {
-                // just a reference to a variable, parameter, or output
-                return new VariableAccessSyntax(identifier);
+                var functionCall = FunctionCallAccess(identifier);
+
+                return new FunctionCallSyntax(
+                    functionCall.Identifier,
+                    functionCall.OpenParen,
+                    functionCall.ArgumentNodes,
+                    functionCall.CloseParen);
             }
+            // returns variable access
+            return new VariableAccessSyntax(identifier);
+        }
 
-            var openParen = this.Expect(TokenType.LeftParen, b => b.ExpectedCharacter("("));
+        /// <summary>
+        /// Method that gets a function call identifier, its arguments plus open and close parens
+        /// </summary>
+        private (IdentifierSyntax Identifier, Token OpenParen, IEnumerable<FunctionArgumentSyntax> ArgumentNodes, Token CloseParen) FunctionCallAccess(IdentifierSyntax functionName)
+        {
+           var openParen = this.Expect(TokenType.LeftParen, b => b.ExpectedCharacter("("));
 
-            var arguments = new List<(SyntaxBase expression, Token? comma)>();
+            var argumentNodes = FunctionCallArguments();
+
+            // consuming right paren, FunctionCallArguments() already validated that it exists
+            var closeParen = this.reader.Read();
```

Comment:
I'd rather just explicitly call `Expect(...)` here than using `reader`, even though it's technically not necessary. It makes the code easier to follow, and a bit more resilient to future bugs.

Reply:
Yeah +1 on that.

---
In reply to: [486679582](https://github.com/Azure/bicep/pull/474#discussion_r486679582) [](ancestors = 486679582)